File: frontend/src/api/intelligence.ts
import { request } from "./client";
import type { TaskStatusResponse } from "./career-analysis";

export interface TaskProgress {
  task_id: string;
  step_index: number;
  total_steps: number;
  step_label: string | null;
  sub_progress: { done: number; total: number } | null;
}

export interface AnalyzeGitHubPayload {
  include_forks?: boolean;
}

export interface PositionScores {
  backend: number;
  frontend: number;
  fullstack: number;
  sre: number;
  cloud: number;
  missing_skills: string[];
}

export interface AnalysisResponse {
  username: string;
  repos_analyzed: number;
  unique_skills: number;
  analyzed_at: string;
  languages: Record<string, number>;
  /** 依存関係から検出したフレームワーク名 → 使用リポジトリ数 */
  detected_frameworks: Record<string, number>;
  /** ルートファイルから検出した DevTools 名 → 使用リポジトリ数 */
  detected_devtools: Record<string, number>;
  /** ルートファイルから検出したインフラツール名 → 使用リポジトリ数 */
  detected_infras: Record<string, number>;
  position_scores: PositionScores | null;
}

export interface CachedAnalysisResponse {
  analysis_result: AnalysisResponse | null;
  position_advice: string | null;
  status?: string;
  error_message?: string;
  error_code?: string;
  /** LLM 不在など、分析自体は完了したが部分的に欠落した場合の警告 */
  warning_message?: string;
}

/**
 * GitHub プロフィールの分析を開始します（202 非同期）。
 */
export function analyzeGitHub(payload: AnalyzeGitHubPayload): Promise<{ status: string }> {
  return request<{ status: string }>("/api/intelligence/analyze", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

/**
 * DB に保存された分析キャッシュを取得します。
 */
export function getAnalysisCache(): Promise<CachedAnalysisResponse> {
  return request<CachedAnalysisResponse>("/api/intelligence/cache");
}

/**
 * 分析ステータスを取得します（ポーリング用）。
 */
export function getAnalysisCacheStatus(): Promise<TaskStatusResponse> {
  return request<TaskStatusResponse>("/api/intelligence/cache/status");
}

/**
 * GitHub 分析タスクの進捗を取得します。
 * Redis が利用できない場合は step_index=0 のデフォルト値が返ります。
 */
export function getAnalysisProgress(): Promise<TaskProgress> {
  return request<TaskProgress>("/api/intelligence/progress");
}

/**
 * 失敗した GitHub 分析タスクを手動で再実行します（202 非同期）。
 */
export function retryAnalyzeGitHub(
  payload: AnalyzeGitHubPayload = {},
): Promise<{ status: string }> {
  return request<{ status: string }>("/api/intelligence/analyze/retry", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

File: backend/app/schemas/intelligence.py
"""キャリアインテリジェンス API 用の Pydantic スキーマ。"""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class AnalyzeRequest(BaseModel):
    include_forks: bool = Field(
        False,
        description="分析にフォークしたリポジトリを含めるかどうか",
    )


class PositionScoresResponse(BaseModel):
    """5軸のエンジニアポジションスコア。"""

    backend: int = Field(0, description="バックエンド適性スコア (0-100)")
    frontend: int = Field(0, description="フロントエンド適性スコア (0-100)")
    fullstack: int = Field(0, description="フルスタック適性スコア (0-100)")
    sre: int = Field(0, description="SRE適性スコア (0-100)")
    cloud: int = Field(0, description="クラウド適性スコア (0-100)")
    missing_skills: List[str] = Field(
        default_factory=list,
        description="フルスタックエンジニアに不足しているスキル",
    )


class AnalysisResponse(BaseModel):
    username: str
    repos_analyzed: int
    unique_skills: int
    analyzed_at: str
    languages: Dict[str, int] = Field(
        default_factory=dict,
        description="言語ごとのバイト数（GitHub linguist ベース）",
    )
    position_scores: Optional[PositionScoresResponse] = Field(
        None,
        description="エンジニアポジションスコア（5軸）",
    )


class PositionAdviceResponse(BaseModel):
    """現状分析+学習アドバイス。"""

    advice: str = Field("", description="LLM による現状分析と学習アドバイス")
    available: bool = Field(True, description="LLM サービスが利用可能かどうか")


class CachedAnalysisResponse(BaseModel):
    """DB に保存された分析結果・学習アドバイスを返す。"""

    analysis_result: Optional[Dict[str, Any]] = None
    position_advice: Optional[str] = None
    status: Optional[str] = None
    error_message: Optional[str] = None
    error_code: Optional[str] = None


class SubProgress(BaseModel):
    """ステップ内の細粒度な進捗（リポジトリ詳細取得ステップ専用）。"""

    done: int
    total: int


class ProgressResponse(BaseModel):
    """GitHub 分析タスクの進捗情報。"""

    task_id: str
    step_index: int = Field(0, description="現在のステップ番号（0 は未開始）")
    total_steps: int = Field(6, description="全ステップ数")
    step_label: Optional[str] = Field(None, description="現在のステップラベル")
    sub_progress: Optional[SubProgress] = Field(
        None, description="リポジトリ詳細取得ステップの細粒度進捗"
    )
